src/category_theory/subobject.lean

```diff
@@ -0,0 +1,906 @@
+def inf_le_left {A : C} (f g : mono_over A) :
+  (inf.obj f).obj g ⟶ f :=
+hom_mk _ rfl
+
+/-- A morphism from the "infimum" of two objects in `mono_over A` to the second object. -/
+def inf_le_right {A : C} (f g : mono_over A) :
+  (inf.obj f).obj g ⟶ g :=
+hom_mk _ pullback.condition
+
+/-- A morphism version of the `le_inf` axiom. -/
+def le_inf {A : C} (f g h : mono_over A) :
+  (h ⟶ f) → (h ⟶ g) → (h ⟶ (inf.obj f).obj g) :=
+begin
+  intros k₁ k₂,
+  refine hom_mk (pullback.lift k₂.left k₁.left _) _,
+  rw [w k₁, w k₂],
+  erw [pullback.lift_snd_assoc, w k₁],
+end
+
+end inf
+
+section sup
+variables [has_images C] [has_binary_coproducts C]
+
+/-- When `[has_images C] [has_binary_coproducts C]`, `mono_over A` has a `sup` construction,
+which is functorial in both arguments,
+and which on `subobject A` will induce a `semilattice_sup`. -/
+def sup  {A : C} : mono_over A ⥤ mono_over A ⥤ mono_over A :=
+curry_obj ((forget A).prod (forget A) ⋙ uncurry.obj over.coprod ⋙ image)
+
+/-- A morphism version of `le_sup_left`. -/
+def le_sup_left {A : C} (f g : mono_over A) :
+  f ⟶ (sup.obj f).obj g :=
+begin
+  refine hom_mk (coprod.inl ≫ factor_thru_image _) _,
+  erw [category.assoc, image.fac, coprod.inl_desc],
+  refl,
+end
+
+/-- A morphism version of `le_sup_right`. -/
+def le_sup_right {A : C} (f g : mono_over A) :
+  g ⟶ (sup.obj f).obj g :=
+begin
+  refine hom_mk (coprod.inr ≫ factor_thru_image _) _,
+  erw [category.assoc, image.fac, coprod.inr_desc],
+  refl,
+end
+
+/-- A morphism version of `sup_le`. -/
+def sup_le {A : C} (f g h : mono_over A) :
+  (f ⟶ h) → (g ⟶ h) → ((sup.obj f).obj g ⟶ h) :=
+begin
+  intros k₁ k₂,
+  refine hom_mk _ _,
+  apply image.lift ⟨_, h.arrow, coprod.desc k₁.left k₂.left, _⟩,
+  { dsimp,
+    ext1,
+    { simp [w k₁] },
+    { simp [w k₂] } },
+  { apply image.lift_fac }
+end
+
+end sup
+
+end mono_over
+
+/-!
+We now construct the subobject lattice for `X : C`,
+as the quotient by isomorphisms of `mono_over X`.
+
+Since `mono_over X` is a thin category, we use `thin_skeleton` to take the quotient.
+
+Essentially all the structure defined above on `mono_over X` descends to `subobject X`,
+with morphisms becoming inequalities, and isomorphisms becoming equations.
+-/
+
+/--
+The category of subobjects of `X : C`, defined as isomorphism classes of monomorphisms into `X`.
+-/
+@[derive [partial_order, category]]
+def subobject (X : C) := thin_skeleton (mono_over X)
+
+namespace subobject
+
+/-- Convenience constructor for a subobject. -/
+abbreviation mk {X A : C} (f : A ⟶ X) [mono f] : subobject X :=
+(to_thin_skeleton _).obj (mono_over.mk' f)
+
+/-- Any functor `mono_over X ⥤ mono_over Y` descends to a functor
+`subobject X ⥤ subobject Y`, because `mono_over Y` is thin. -/
+def lower {Y : D} (F : mono_over X ⥤ mono_over Y) : subobject X ⥤ subobject Y :=
+thin_skeleton.map F
+
+/-- Isomorphism functors become (evil!) equal when lowered to `subobject`. -/
```

Comment:
```suggestion
/-- Isomorphism functors become equal when lowered to `subobject`.
(It's not as evil as usual to talk about equality between functors because the categories are thin and skeletal.-/
```